pkg/authorization/admission/restrictusers/subjectchecker_test.go

```diff
@@ -0,0 +1,273 @@
+		{
+			name: "prohibit regular user when another user matches on label selector",
+			checker: mustNewSubjectChecker(t,
+				&authorizationapi.RoleBindingRestrictionSpec{
+					UserRestriction: &authorizationapi.UserRestriction{
+						Selectors: []unversioned.LabelSelector{
+							{MatchLabels: map[string]string{"foo": "bar"}},
+						},
+					},
+				}),
+			subject:     userBobRef,
+			shouldAllow: false,
+		},
+		{
+			name: "allow regular group by literal name match",
+			checker: mustNewSubjectChecker(t,
+				&authorizationapi.RoleBindingRestrictionSpec{
+					GroupRestriction: &authorizationapi.GroupRestriction{
+						Groups: []string{groupRef.Name},
+					},
+				}),
+			subject:     groupRef,
+			shouldAllow: true,
+		},
+		{
+			name: "allow regular group by label selector match",
+			checker: mustNewSubjectChecker(t,
+				&authorizationapi.RoleBindingRestrictionSpec{
+					GroupRestriction: &authorizationapi.GroupRestriction{
+						Selectors: []unversioned.LabelSelector{
+							{MatchLabels: map[string]string{"baz": "quux"}},
+						},
+					},
+				}),
+			subject:     groupRef,
+			shouldAllow: true,
+		},
+		{
+			name: "allow service account by literal name match",
+			checker: mustNewSubjectChecker(t,
+				&authorizationapi.RoleBindingRestrictionSpec{
+					ServiceAccountRestriction: &authorizationapi.ServiceAccountRestriction{
+						ServiceAccounts: []authorizationapi.ServiceAccountReference{
+							{
+								Name:      serviceaccountRef.Name,
+								Namespace: serviceaccountRef.Namespace,
+							},
+						},
+					},
+				}),
+			subject:     serviceaccountRef,
+			shouldAllow: true,
+		},
+		{
+			name: "allow service account by literal name match with implicit namespace",
+			checker: mustNewSubjectChecker(t,
+				&authorizationapi.RoleBindingRestrictionSpec{
+					ServiceAccountRestriction: &authorizationapi.ServiceAccountRestriction{
+						ServiceAccounts: []authorizationapi.ServiceAccountReference{
+							{Name: serviceaccountRef.Name},
+						},
+					},
+				}),
+			subject:     serviceaccountRef,
+			shouldAllow: true,
+		},
+		{
+			name: "allow service account by match on namespace",
+			checker: mustNewSubjectChecker(t,
+				&authorizationapi.RoleBindingRestrictionSpec{
+					ServiceAccountRestriction: &authorizationapi.ServiceAccountRestriction{
+						Namespaces: []string{serviceaccountRef.Namespace},
+					},
+				}),
+			subject:     serviceaccountRef,
+			shouldAllow: true,
+		},
+	}
+
+	kclient := fake.NewSimpleClientset(objects...)
+	oclient := otestclient.NewSimpleFake(objects...)
+	groupCache := usercache.NewGroupCache(&groupCache{[]userapi.Group{group}})
+	groupCache.Run()
+	// This is a terrible, horrible, no-good, very bad hack to avoid a race
+	// condition between the test "allow regular user by group membership"
+	// and the group cache's initialisation.
+	for {
+		if groups, _ := groupCache.GroupsFor(group.Users[0]); len(groups) == 1 {
+			break
+		}
+		time.Sleep(time.Millisecond)
+	}
```

Comment:
so the sleep is good enough for now, but what you probably really wanted was to create an interface for the groupcache and implement your own mock for the test.